internal/serve/middleware/middleware.go

```diff
@@ -177,6 +179,70 @@ func (c cspItem) String() string {
 	return fmt.Sprintf("%s %s;", c.ContentType, strings.Join(c.Policy, " "))
 }
 
+// LoggingMiddleware is a middleware that logs requests to the logger.
+func LoggingMiddleware() func(http.Handler) http.Handler {
+	return func(next http.Handler) http.Handler {
+		return http.HandlerFunc(func(rw http.ResponseWriter, req *http.Request) {
+			mw := mutil.WrapWriter(rw)
+
+			reqCtx := req.Context()
+			logFields := log.F{
+				"req": middleware.GetReqID(reqCtx),
+			}
+			logCtx := log.Set(reqCtx, log.Ctx(reqCtx).WithFields(logFields))
+
+			tenant, err := tenant.GetTenantFromContext(reqCtx)
+			if err != nil {
+				log.Ctx(logCtx).Info("tenant cannot be derived from context")
+			}
+			if tenant != nil {
+				logFields["tenant_name"] = tenant.Name
+				logFields["tenant_id"] = tenant.ID
+				logCtx = log.Set(reqCtx, log.Ctx(reqCtx).WithFields(logFields))
+			}
+
+			logRequestStart(logCtx, req)
+			started := time.Now()
+			req = req.WithContext(logCtx)
+
+			next.ServeHTTP(mw, req)
+			ended := time.Since(started)
+			logRequestEnd(logCtx, req, mw, ended)
+		})
+	}
+}
+
+func logRequestStart(ctx context.Context, req *http.Request) {
+	l := log.Ctx(ctx).WithFields(
+		log.F{
+			"subsys":    "htttp",
+			"path":      req.URL.String(),
+			"method":    req.Method,
+			"ip":        req.RemoteAddr,
+			"host":      req.Host,
+			"useragent": req.Header.Get("User-Agent"),
+		},
+	)
+
+	l.Info("starting request")
+}
+
+func logRequestEnd(ctx context.Context, req *http.Request, mw mutil.WriterProxy, duration time.Duration) {
+	l := log.Ctx(ctx).WithFields(log.F{
+		"subsys":   "http",
+		"path":     req.URL.String(),
+		"method":   req.Method,
+		"status":   mw.Status(),
+		"bytes":    mw.BytesWritten(),
+		"duration": duration,
+	})
+	if routeContext := chi.RouteContext(req.Context()); routeContext != nil {
```

Comment:
Since you're updating the request with the context in line 206, you don't need to pass the ctx as a parameter and can use the one from the request here (line 231, by using: `log.Ctx(req.Context())`.

Please notice you're already using that approach in line 239.

Reply:
yeah, makes sense! I'm using logCtx here which is already the request context.